Data/ByteString.hs

```diff
@@ -389,6 +389,11 @@ snoc (BS x l) c = unsafeCreate (l+1) $ \p -> unsafeWithForeignPtr x $ \f -> do
 
 -- | /O(1)/ Extract the first element of a ByteString, which must be non-empty.
 -- An exception will be thrown in the case of an empty ByteString.
+--
+-- This function is both partial and slow.
+-- Please consider using either 'uncons', which is total, or, if you are
+-- adamant that the argument is non-empty, 'Data.ByteString.Unsafe.unsafeHead',
+-- which is branchless and fast.
```

Comment:
This reads like we want users to pick either `uncons` or `unsafeHead` instead of `head`. Is this your goal?

How about something like this:

```suggestion
-- Related functions:
--
--   * 'uncons', for safe access to the 'head' and 'tail' of a ByteString
--   * 'Data.ByteString.Unsafe.unsafeHead' 
```

Reply:
> Is this your goal?

To be honest yes, I'd like to see things rolling away from `head`. What's your take about it?

Reply:
I think users ought to be careful with `head`. But there still are legitimate use cases for it, where the code using `head` is preceded by a length check on the bytestring.

So I'm fine with warning about the partiality, and pointing out the alternatives.

But I don't want to push users to use `unsafeHead` rather than `head`. I think it's too dangerous to be used outside of the most performance-critical code.

Reply:
OK, let's make it less controversial. Updated.